commit 89d8346797e106dd52f02c435232b8668e2e0b98 (refs/changes/26/3926/3)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-05-15T16:35:46-07:00 (1 year, 5 months ago)
    
    TRITON-375 node-moray should expose similar metrics to moray server

diff --git a/CHANGES.md b/CHANGES.md
index 1801bb9..9365f96 100644
--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,5 +1,9 @@
 # Changelog
 
+## v3.5.0
+
+* [TRITON-375](http://smartos.org/bugview/TRITON-375) node-moray should expose similar metrics to moray server
+
 ## v3.4.2
 
 * [MORAY-455](http://smartos.org/bugview/MORAY-455) moray asks fast to hold onto too many recent RPC requests
diff --git a/bin/getobject b/bin/getobject
index b5b6e89..891627c 100755
--- a/bin/getobject
+++ b/bin/getobject
@@ -7,15 +7,17 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
+var artedi = require('artedi');
 var cmdutil = require('cmdutil');
 var moray = require('../lib');
 var moraycli = require('../lib/cmd');
 
 var clientOptions, parser, client;
 var bucket, key, getOptions;
+var collector;
 var prettyPrint = true;
 
 cmdutil.configure({
@@ -61,6 +63,13 @@ if ((parser.optind() + 1) >= process.argv.length)
     cmdutil.usage('missing required arguments: "key"');
 key = process.argv[parser.optind() + 1];
 
+if (process.env.DUMP_ARTEDI_METRICS) {
+    collector = artedi.createCollector({labels: {
+        service: 'getobject'
+    }});
+    clientOptions.collector = collector;
+}
+
 client = moray.createClient(clientOptions);
 client.on('error', cmdutil.fail);
 client.on('connect', function onConnect() {
@@ -75,6 +84,18 @@ client.on('connect', function onConnect() {
             console.log(JSON.stringify(obj));
         }
 
+        if (collector) {
+            collector.collect(artedi.FMT_PROM,
+                function _outputMetrics(err, metrics) {
+                    if (err) {
+                        cmdutil.fail(err);
+                        return;
+                    }
+                    console.error('[artedi metrics]');
+                    console.error(metrics);
+                });
+        }
+
         client.close();
     });
 });
diff --git a/docs/man/man3/moray.md b/docs/man/man3/moray.md
index 4ddab86..1e4b95e 100644
--- a/docs/man/man3/moray.md
+++ b/docs/man/man3/moray.md
@@ -1,4 +1,4 @@
-# moray 3 "January 2017" Moray "Moray Client Library"
+# moray 3 "May 2018" Moray "Moray Client Library"
 
 ## NAME
 
@@ -205,6 +205,11 @@ All constructor invocations must also provide one of the following:
 
 Callers may also provide:
 
+`collector` (object)
+  An [artedi](https://github.com/joyent/node-artedi) metrics collector. This
+  will be passed to the underlying node-fast component which will add metrics
+  for count and duration of RPC calls.
+
 `cueballOptions` (object)
   Overrides cueball-related options, including various timeouts and delays.
   For specific options that can be overridden here, see the source.  **NOTE:
diff --git a/lib/client.js b/lib/client.js
index 1a2a6da..c2c08f0 100644
--- a/lib/client.js
+++ b/lib/client.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2017, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -81,6 +81,7 @@ function MorayClient(options) {
     EventEmitter.call(this);
 
     assert.object(options, 'options');
+    assert.optionalObject(options.collector, 'options.collector');
     assert.object(options.log, 'options.log');
     assert.optionalBool(options.unwrapErrors, 'options.unwrapErrors');
     assert.optionalBool(options.failFast, 'options.failFast');
@@ -108,6 +109,11 @@ function MorayClient(options) {
 
     this.log.debug(coptions, 'init');
 
+    /* Optional artedi metrics collector that we'll pass to fast, if set. */
+    if (options.collector) {
+        this.collector = options.collector;
+    }
+
     if (coptions.mode == 'srv') {
         resolverInput = cueballOptions.domain;
     } else {
@@ -327,6 +333,7 @@ MorayClient.prototype.createFastConnection =
 
     return (new FastConnection({
         'address': backend.address,
+        'collector': this.collector,
         'port': backend.port,
         'nRecentRequests': fastNRecentRequests,
         'tcpKeepAliveInitialDelay': dflClientTcpKeepAliveIdle,
diff --git a/lib/fast_connection.js b/lib/fast_connection.js
index d9d8781..8b57811 100644
--- a/lib/fast_connection.js
+++ b/lib/fast_connection.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2016, Joyent, Inc.
+ * Copyright (c) 2018, Joyent, Inc.
  */
 
 /*
@@ -28,6 +28,9 @@ module.exports = FastConnection;
  *     address                  IPv4 or IPv6 address, interpreted by
  *     (string)                 net.createConnection().
  *
+ *     collector                Artedi metric collector, used by node-fast to
+ *     (object)                 provide fast client metrics. [optional]
+ *
  *     port                     TCP port, used for net.createConnection().
  *     (number)
  *
@@ -43,6 +46,7 @@ module.exports = FastConnection;
 function FastConnection(args) {
     assert.object(args, 'args');
     assert.string(args.address, 'args.address');
+    assert.optionalObject(args.collector, 'args.collector');
     assert.number(args.port, 'args.port');
     assert.object(args.log, 'args.log');
     assert.number(args.nRecentRequests, 'args.nRecentRequests');
@@ -60,9 +64,10 @@ function FastConnection(args) {
     this.fc_sock = net.createConnection(args.port, args.address);
     this.fc_destroyed = false;
     this.fc_fast = new fast.FastClient({
+        'collector': args.collector, /* Optional: artedi metrics collector */
+        'log': args.log,
         'nRecentRequests': args.nRecentRequests,
-        'transport': this.fc_sock,
-        'log': args.log
+        'transport': this.fc_sock
     });
 
     /*
diff --git a/package.json b/package.json
index 4f4282c..276aba8 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
     "name": "moray",
     "description": "Moray client library",
-    "version": "3.4.2",
+    "version": "3.5.0",
     "author": "Joyent (joyent.com)",
     "keywords": [ "moray" ],
     "main": "./lib/index.js",
@@ -10,13 +10,14 @@
         "url": "git+ssh://git@github.com:joyent/node-moray.git"
     },
     "dependencies": {
+        "artedi": "1.3.0",
         "assert-plus": "^1.0.0",
         "backoff": "^2.4.1",
         "bunyan": "^1.3.2",
         "cmdutil": "^1.1.0",
         "cueball": "^2.3.0",
         "extsprintf": "^1.3.0",
-        "fast": "^2.1.0",
+        "fast": "^2.5.0",
         "libuuid": "0.2.1",
         "jsprim": "^1.3.0",
         "posix-getopt": "^1.0.0",
